diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 83a6b68d8..4bba21f21 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -3005,6 +3005,15 @@ choice
 		bool "Extend builtin kernel arguments with bootloader arguments"
 endchoice
 
+config MIPS_TUX64
+	bool "Tux64 support"
+	depends on MACH_NINTENDO64
+	default n
+	help
+	  Enables support for Tux64 exclusive features, such as extended boot
+	  parameters. This introduces breaking changes for existing bootloaders.
+
+	  If unsure, say n.
 endmenu
 
 config LOCKDEP_SUPPORT
