diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 83a6b68d8..f22b67850 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -3005,8 +3005,28 @@ choice
 		bool "Extend builtin kernel arguments with bootloader arguments"
 endchoice
 
+config MIPS_TUX64
+	bool "Tux64 support"
+	depends on MACH_NINTENDO64
+	default n
+	help
+	  Enables support for Tux64 exclusive features, such as extended boot
+	  parameters. This introduces breaking changes for existing bootloaders.
+
+	  If unsure, say n.
+
+config TUX64_PROC_FS
+	bool "Provide N64 system information via /proc/n64"
+	depends on MIPS_TUX64 && PROC_FS
+	default y
+	help
+	  Provides system information such as console type, boot information, and
+	  hardware revision to userspace applications via /proc/n64/.
+
 endmenu
 
+
+
 config LOCKDEP_SUPPORT
 	bool
 	default y
